Shark Cage/createACLinNewDesktop/createACLinNewDesktop/util.cpp

```diff
@@ -4,121 +4,121 @@
 void enumerateSidsAndHashes(PTOKEN_ACCESS_INFORMATION pToken);
 ULONG getCurrentSessionID();
 void deconstructToken(tokenStructures &tokenDeconstructed, HANDLE &userToken);
+bool changeTokenCreationPrivilege(bool privilegeStatus);
+bool addGroupToTokenGroups(PSID sid, tokenStructures &tokenDeconstructed);
 
 
 
-namespace util {
-
-	LPWSTR  group_name = L"Dummy_group"; //TODO: this should be constant, fix the conversions
-
-
-	//this is all obsolete now  - keeping this only for grading purposes
-	//reason: there is no way to arbitrarilly add a group to the token
-	//we are going to create a new token from scratch using NtCreateToken
-	bool getModifiedToken(PSID sid, HANDLE &token) {
-
+namespace tokenLib {
 
+	bool createLocalGroup(LPWSTR groupName, PSID &sid){
+		SID_NAME_USE accountType;
 		LOCALGROUP_INFO_0 localgroup_info;
-		LPTSTR   pUser_name = NULL;
-		HANDLE current_process_handle;
-		HANDLE user_token_h;
-		DWORD bufferSize = 0;
-
+		localgroup_info.lgrpi0_name = groupName;
 
-		//create a dummy group
-		localgroup_info.lgrpi0_name = group_name;
-		if (NetLocalGroupAdd(NULL, 0, (LPBYTE)&localgroup_info, NULL) != NERR_Success)
-			return emergencyExit(pUser_name);
+		NetLocalGroupDel(NULL, groupName);
 
+		if (NetLocalGroupAdd(NULL, 0, (LPBYTE)&localgroup_info, NULL) != NERR_Success) {
+			sid = NULL;
+			return false;
+		}
 
-		//add user to that group
-		GetUserName(NULL, &bufferSize);
-		pUser_name = (LPTSTR) new BYTE[bufferSize];
-		GetUserName(pUser_name, &bufferSize);
-		LOCALGROUP_MEMBERS_INFO_3 localgroup_member_info;
-		localgroup_member_info.lgrmi3_domainandname = pUser_name;
-		if (NetLocalGroupAddMembers(NULL, group_name, 3, (LPBYTE)&localgroup_member_info, 1) != NERR_Success)
-			return emergencyExit(pUser_name);
-
-		//get user token (containig dummy group) - TODO: I need handle of a different process this one does not have the priviledge
-		current_process_handle = GetCurrentProcess();
-		if (!OpenProcessToken(current_process_handle, TOKEN_ALL_ACCESS, &user_token_h))
-			return emergencyExit(pUser_name);
-
-		bufferSize = 0;
-		GetTokenInformation(user_token_h, TokenAccessInformation, NULL, 0, &bufferSize);
-		PTOKEN_ACCESS_INFORMATION pTokenUser = (PTOKEN_ACCESS_INFORMATION) new BYTE[bufferSize];
-		memset(pTokenUser, 0, bufferSize);
-		GetTokenInformation(user_token_h, TokenAccessInformation, pTokenUser, bufferSize, &bufferSize);
-
-		enumerateSidsAndHashes(pTokenUser);
-		//replace the dummy group in token with actual group - requires to know the sid of the dummy_group
-		//change the hashes - how - not possible  - one hash for all of the tokens - aborting this approach
-
-
-		//delete dummy group
-		NetLocalGroupDel(NULL, group_name);
-
-		//assign token to handle
+		DWORD bufferSize = 0, buffer2Size = 0;
 
+		LookupAccountName(NULL, groupName, NULL, &bufferSize, NULL, &buffer2Size, &accountType);
+		sid = (PSID) new BYTE[bufferSize];
+		LPTSTR domain = (LPTSTR) new BYTE[buffer2Size];

```

Comment:
this and the line above is a memory leak, isn't it? at least `domain` can be free'd locally. for the `SID` maybe a smart pointer could be used, another possibility would be to clean up all created resources returned to the caller in the respective clean function (`deleteLocalGroup` in this case)

Reply:
You are right, this is a memmory leak... and not the only one... for example whole "tokenDeconstructed" structure is not being deallocated becuase I am not sure how is that affecting the new token. I´ve created a PR the moment I´ve reached a functional state with a stable API so you guys can start integrating it into the project. I will continue perfecting the implementation in parallel (should not cause problem if I keep the behaviour unchanged)

Reply:
In my opinion it is dangerous to have the caller do clean-up duty of any kind. There are several ways to allocate memory in C++ and you have to use the matching deallocation scheme to not corrupt the memory of your program. Adding it to the documentation might be fine for us / this project but as a general approach to API design there are better alternatives

Reply:
All right, I get your point, what I was trying to drive forward, was that this is not that unusual in this corner of WINAPI. Have a look at [ConvertSidToStringSidA](https://docs.microsoft.com/sk-sk/windows/desktop/api/sddl/nf-sddl-convertsidtostringsida). In documentation it states 

> A pointer to a variable that receives a pointer to a null-terminated SID string. To free the returned buffer, call the `LocalFree` function.

And yeah... I know that WINAPI is not a good example for a proper API design 😬 I just conformed to id... anyway I will rethink it somehow in the near future, stay tuned 🍿